### PR TITLE
debug tools

### PR DESCRIPTION
**Pull request by MCJOHN974**
Added some debug print in compiler and small script for wasm disassemble. @shishkin-pavel maybe it will be useful for you

### CODE REVIEW

File: bin/evm2near/src/compile.rs [hunk 1]
```diff
@@ -252,11 +255,12 @@ impl Compiler {
                     match block.origin {
                         CaterpillarLabel::Original(orig_label) => {
                             let code = &program.0[orig_label.code_start..orig_label.code_end];
-                            for op in code {
+                            for (idx, op) in code.into_iter().enumerate() {
```
Comment (shishkin-pavel) [inline bin/evm2near/src/compile.rs:258]: just `iter` instead of `into_iter`

File: bin/evm2near/src/compile.rs [hunk 2]
```diff
@@ -274,21 +278,87 @@ impl Compiler {
...
     ) {
         assert_ne!(self.evm_start_function, 0); // filled in during emit_start()
         assert_eq!(self.evm_exec_function, 0); // filled in below
 
         let mut wasm: Vec<Instruction> = Default::default();
-        self.unfold_cfg(input_program, input_cfg, &mut wasm);
+        let debug_info = self.unfold_cfg(input_program, input_cfg, &mut wasm);
         wasm.push(Instruction::End);
 
+        id2offs.insert(usize::MAX, usize::MAX);
```
Comment (shishkin-pavel) [inline bin/evm2near/src/compile.rs:298]: ? =)

File: bin/evm2near/src/compile.rs [hunk 3]
```diff
@@ -274,21 +278,87 @@ impl Compiler {
...
+        // debug
+
+        for (wasmid, evmid) in &debug_info {
+            println!("wasmid {} => evmid {}", wasmid, evmid);
+        }
+
+        let wasmline2offs = |line: usize| {
```
Comment (shishkin-pavel) [inline bin/evm2near/src/compile.rs:305]: `.map` on option instead

File: bin/evm2near/src/compile.rs [hunk 4]
```diff
@@ -274,21 +278,87 @@ impl Compiler {
...
+            wasm.clone()
+                .into_iter()
+                .enumerate()
+                .map(|(idx, instr)| {
+                    if instr == Instruction::Else || instr == Instruction::End {
+                        for _ in 0..2 {
+                            shift.pop();
+                        }
+                    }
+                    let res: String;
```
Comment (shishkin-pavel) [inline bin/evm2near/src/compile.rs:327]: `.map`

**Review by shishkin-pavel — COMMENTED** [review]
we definitely should avoid committing debug print lines with usual println
also, im not happy about fs::write calls everywhere not using specific config option

**Comment by MCJOHN974** [comment]
> 

You mean we need use smth like ?
```
if congig.debug {
  std::fs::...
  }
 }
 or smth like
 ```
 debug_log!()
 ```
 ?

**Comment by shishkin-pavel** [comment]
> You mean we need use smth like ?

separate debugging macro will be a good choice for current `eprintln!`-like things

Also, Im thinking about separate compilation option (+ function call?) for different compilation stages logs
for example, currently we have several levels of cfg trensformation and probably it would be nice to have a key for dumping results of all that stages in directory (provided as compilation option) as different of files
but probably it is not useful enough to think about that right now. lets just fix all the issues

**Comment by MCJOHN974** [comment]
@shishkin-pavel  I think I fix all current problems, please take a look

**Comment by shishkin-pavel** [comment]
that branch is already contained in #5 as faa3d26 (squashed)

**Pull request closed without merging**